File: 22.linq-workshop-kata/ts/tests/01-fundamentals/basics.test.ts
import { describe, expect, it } from "vitest";
import { Angle } from "../../src/01-fundamentals/angle";
import { isEven, isPalindrome, toAngle } from "../../src/01-fundamentals/basics";

describe("Basics", () => {
  it("palindrome", () => {
    expect(isPalindrome("racecar")).toBe(true);
    expect(isPalindrome("kayak")).toBe(true);
    expect(isPalindrome("nein")).toBe(false);
  });

  it("evenNumbers", () => {
    expect(isEven(2)).toBe(true);
    expect(isEven(101)).toBe(false);
  });

  it("angles", () => {
    expect(toAngle(1)).toEqual(new Angle(1));
    expect(toAngle(361)).toEqual(new Angle(1));
    expect(toAngle(360)).toEqual(new Angle(0));
    expect(toAngle(-1)).toEqual(new Angle(-1));
  });
});

File: 22.linq-workshop-kata/java/src/test/java/linq_workshop/fundamentals/BasicsTest.java
package linq_workshop.fundamentals;

import org.junit.jupiter.api.Test;

import static linq_workshop.fundamentals.Basics.isEven;
import static linq_workshop.fundamentals.Basics.isPalindrome;
import static linq_workshop.fundamentals.Basics.toAngle;
import static org.assertj.core.api.Assertions.assertThat;

class BasicsTest {

    @Test
    void palindrome() {
        assertThat(isPalindrome("racecar")).isTrue();
        assertThat(isPalindrome("kayak")).isTrue();
        assertThat(isPalindrome("nein")).isFalse();
    }

    @Test
    void evenNumbers() {
        assertThat(isEven(2)).isTrue();
        assertThat(isEven(101)).isFalse();
    }

    @Test
    void angles() {
        assertThat(toAngle(1)).isEqualTo(new Angle(1));
        assertThat(toAngle(361)).isEqualTo(new Angle(1));
        assertThat(toAngle(360)).isEqualTo(new Angle(0));
        assertThat(toAngle(-1)).isEqualTo(new Angle(-1));
    }
}

final class Basics {

    private Basics() {
    }

    static boolean isPalindrome(String value) {
        throw new UnsupportedOperationException();
    }

    static boolean isEven(int number) {
        throw new UnsupportedOperationException();
    }

    static Angle toAngle(int degrees) {
        throw new UnsupportedOperationException();
    }
}

record Angle(int degrees) {

    Angle {
        degrees = normalize(degrees);
    }

    private static int normalize(int value) {
        return (value % 360 + 360) % 360;
    }

    Angle plus(Angle other) {
        return new Angle(degrees + other.degrees);
    }

    Angle minus(Angle other) {
        return new Angle(degrees - other.degrees);
    }

    @Override
    public String toString() {
        return degrees + "°";
    }
}
